packages/blocks/src/api/constants.js

```diff
@@ -118,6 +118,26 @@ export const __EXPERIMENTAL_STYLE_PROPERTY = {
 		value: [ 'spacing', 'blockGap' ],
 		support: [ 'spacing', 'blockGap' ],
 	},
+	'--wp--style--padding-top': {
+		value: [ 'spacing', 'padding', 'top' ],
+		support: [ 'spacing', 'padding' ],
+		rootOnly: true,
```

Comment:
This feels super hacky, but without it these styles get applied to all blocks that have padding set on them. Any ideas welcome!

Reply:
Should global styles styles also define these vars for specific blocks?

I bumped the padding value of the Group block in Global styles and I see this in the `global-styles-inline-css` style tag:

```css
body {
	...
	--wp--style--padding-top: 173px;
	--wp--style--padding-right: 173px;
	--wp--style--padding-bottom: 173px;
	--wp--style--padding-left: 173px;
	--wp--style--block-gap: 1.5rem;
}
...
.wp-block-group {
	--wp--style--padding-top: 142px;
	--wp--style--padding-right: 142px;
	--wp--style--padding-bottom: 142px;
	--wp--style--padding-left: 142px;
}
```

The editor reflects the changes but the frontend doesn't, because, I guess, we'd need to either update the blocks' CSS to use these vars or print it on the fly somewhere in the backend.

On a related note, gap also gets special treatment in [WP_Theme_JSON_Gutenberg](https://github.com/WordPress/gutenberg/blob/848db705d6436bc267234910eb653968e6bf2893/lib/compat/wordpress-5.9/class-wp-theme-json-5-9.php#L303) via  `'blockGap' => 'top',`, the effect of which is that `--wp--style--block-gap` is only rendered at the "top" (or root) level in global styles. Effectively the body tag.

But we also want block-level global styles, when set, to override any root value. It's something we're trying resolve over at https://github.com/WordPress/gutenberg/pull/39870

It seems like we do need some sort of flag that says _"Hey, this property only lives at the root level, so output its style rule in the body tag and nowhere else"_.

I'm not sure what that would look like yet, maybe a `const VALID_TOP_LEVEL_STYLES` in `WP_Theme_JSON_Gutenberg` that explicitly defines root-level-only styles 🤷

Reply:
> Should global styles also output these vars for specific blocks?

Oooooh no it shouldn't! It's actually outputting the variables instead of the properties in the front end now, and that's messing with some of the negative margin calculations 😫 

> It seems like we do need some sort of flag that says "Hey, this property only lives at the root level, so output its style rule in the body tag and nowhere else".

Yup, we definitely need something like this. I was playing around with it on the editor side here, not realising it would also be needed on the server side.

> I'm not sure what that would look like yet, maybe a const VALID_TOP_LEVEL_STYLES in WP_Theme_JSON_Gutenberg that explicitly defines root-level-only styles

Can we make do with the "top" value in `VALID_STYLES` solution we have currently? What shortcomings do you see with it?

Reply:
Ok I made some changes that should fix the block-level global styles issue. 

The thing with padding is that we only need it to be a variable at the root level, so we can easily apply it to all the places we need it: as padding on full-width root level blocks, or as negative margin on their full-width children. I'm not sure if block gap works exactly like that? (I only read through the code of #39870 today, but will look at it more in-depth on Monday).

Reply:
> The thing with padding is that we only need it to be a variable at the root level, so we can easily apply it to all the places we need it: as padding on full-width root level blocks, or as negative margin on their full-width children. I'm not sure if block gap works exactly like that? 

Block gap is a smidge different, for sure.

Support for gap in non root-level global styles was removed, and in `sanitized()` it's being excluded from the non-root level blocks and elements in the merged theme json.  That means that, right now, there's no need to worry about whether the styles will be correctly output.

The side-effect of that removal was that block-level global styles cannot define gap values; this side-effect was, I believe, unintended. This is the "bug" that https://github.com/WordPress/gutenberg/pull/39870 is trying to address.

This is a long-winded way of saying that I'd like to take block gap in the same direction you're taking padding 😄 so I _think_ there might be similarities...

Reply:
Conceptually comparing the root padding to block gap makes sense for the reasons Isabel outlines. Just as an example, we'll need to apply this root padding inside the modal navigation overlay as well, so the overlay padding is the same as the root padding.

Reply:
Oh, thanks for the reminder @jasmussen ! I'd forgotten about the overlay special case. Do you know of anywhere else where we might need to apply root padding?

Reply:
There are many cases where we might want the site padding to match other parts of the site - for example the gaps between navigation items, template parts and galleries. This is an important consideration in theme design - if these spaces all defaulted to something that matched it would mean that designs would look good without needing to tweak all these things individually as we do today cc @kjellr @beafialho

Reply:
> There are many cases where we might want the site padding to match other parts of the site - for example the gaps between navigation items, template parts and galleries etc.

That would be a theme level decision though, wouldn't it? I'd imagine the theme would set a global spacing variable and use it to set both root and block padding (or gap, or whatever else) in theme.json and in the theme stylesheet.

The root padding we're dealing with here is a [top-level style](https://developer.wordpress.org/block-editor/how-to-guides/themes/theme-json/#top-level-styles) so it's only attached to the body. This PR changes the output from a straightforward padding declaration to a custom property, mainly so we can use its value to offset full-width blocks.

Reply:
Yes, that's a theme level decision, but I think its the kind of thing that having a good default for would really help - for example if this setting defaulted to `--wp--style--block-gap` then all these spacing elements would match by default.

Reply:
Oh, I see! 

I changed the default values [here](https://github.com/WordPress/gutenberg/blob/044f281ff5c4907fdcbfeea5d9225e2962e9aada/lib/compat/wordpress-6.0/theme.json#L245) to use the block gap variable - does that work better?